// @nx/nuxt/src/generators/application/lib/ensure-dependencies.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ensureDependencies = ensureDependencies;
const devkit_1 = require("@nx/devkit");
const vue_1 = require("@nx/vue");
const versions_1 = require("../../../utils/versions");
function ensureDependencies(host, options) {
    const devDependencies = {
        '@nx/vite': versions_1.nxVersion, // needed for the nxViteTsPaths plugin and @nx/vite/plugin
        '@nuxt/devtools': versions_1.nuxtDevtoolsVersion,
        '@nuxt/kit': versions_1.nuxtVersion,
        '@nuxt/ui-templates': versions_1.nuxtUiTemplatesVersion,
        nuxt: versions_1.nuxtVersion,
        h3: versions_1.h3Version,
        vue: vue_1.vueVersion,
        'vue-router': vue_1.vueRouterVersion,
        'vue-tsc': vue_1.vueTscVersion,
    };
    if (options.style === 'scss') {
        devDependencies['sass'] = vue_1.sassVersion;
    }
    else if (options.style === 'less') {
        devDependencies['less'] = vue_1.lessVersion;
    }
    return (0, devkit_1.addDependenciesToPackageJson)(host, {}, devDependencies);
}
